libyul/optimiser/UnusedFunctionsCommon.h

```diff
@@ -0,0 +1,172 @@
+{
+	// We skip the function body if it
+	// 1. is empty, or
+	// 2. is a single statement that is an assignment statement whose value is a non-builtin
+	//    function call, or
+	// 3. is a single expression-statement that is a non-builtin function call.
+	// The above cases are simple enough so that the inliner alone can remove the parameters.
+	if (_f.body.statements.empty())
+		return true;
+	if (_f.body.statements.size() == 1)
+	{
+		Statement const& e = _f.body.statements[0];
+		if (std::holds_alternative<Assignment>(e))
+		{
+			if (std::holds_alternative<FunctionCall>(*std::get<Assignment>(e).value))
+			{
+				FunctionCall c = std::get<FunctionCall>(*std::get<Assignment>(e).value);
+				if (!_dialect.builtin(c.functionName.name))
+					return true;
+			}
+		}
+		else if (std::holds_alternative<ExpressionStatement>(e))
+			if (std::holds_alternative<FunctionCall>(std::get<ExpressionStatement>(e).expression))
+			{
+				FunctionCall c = std::get<FunctionCall>(std::get<ExpressionStatement>(e).expression);
+				if (!_dialect.builtin(c.functionName.name))
+					return true;
+			}
+	}
+	return false;
+}
+
+FunctionDefinition createReplacement(
+	FunctionDefinition& _old,
+	std::optional<std::vector<bool>> const& _usedParameters,
+	std::optional<std::vector<bool>> const& _usedReturnVariables,
+	NameDispenser& _nameDispenser,
+	YulString const& _newName)
+{
+	auto generateName = [&](TypedName t)
+	{
+		return TypedName{
+			t.location,
+			_nameDispenser.newName(t.name),
+			t.type
+		};
+	};
+
+	langutil::SourceLocation loc = _old.location;
+	TypedNameList functionParameters;
+	TypedNameList returnVariables;
+	TypedNameList renamedParameters = util::applyMap(_old.parameters, generateName);
+	TypedNameList reducedRenamedParameters;
+	TypedNameList renamedReturnVariables = util::applyMap(_old.returnVariables, generateName);
+	TypedNameList reducedRenamedReturnVariables;
+
+	if (_usedParameters)
+	{
+		std::vector<bool> const& mask = _usedParameters.value();
+		functionParameters = applyBooleanMask(_old.parameters, mask);
+		reducedRenamedParameters = applyBooleanMask(renamedParameters, mask);
+	}
+	else
+	{
+		functionParameters = _old.parameters;
+		reducedRenamedParameters = renamedParameters;
+	}
+
+	if (_usedReturnVariables)
+	{
+		std::vector<bool> const& mask = _usedReturnVariables.value();
+		returnVariables = applyBooleanMask(_old.returnVariables, mask);
+		reducedRenamedReturnVariables = applyBooleanMask(renamedReturnVariables, mask);
+	}
+	else
+	{
+		returnVariables = _old.returnVariables;
+		reducedRenamedReturnVariables = renamedReturnVariables;
+	}
+
+	FunctionDefinition newFunction{
+		loc,
+		_old.name,
+		renamedParameters,
+		reducedRenamedParameters,
+		{loc, {}} // body
+	};
+
+	_old.name = _newName;
+	_old.parameters = std::move(functionParameters);
+	_old.returnVariables = std::move(returnVariables);
+
+	std::swap(newFunction.parameters, renamedParameters);
```

Comment:
Why do you swap here? Shouldn't one be a copy of the other?

Reply:
Yes. This is redundant. Sorry about that, after multiple changes, this is not so cohesive. I'll try to avoid this problem in the future.